### PR TITLE
fix(dir-filter): look for first parent project

### PR DESCRIPTION
**Pull request by Dynge**
In order to see if the directory should be scanned we need to check upward for the first project (assuming that no file can be contained within two or more projects at the same time).

Furthermore Ive changed some variable names and added a log if the directory is ignored because it is determined not to be from the solution project.

### CODE REVIEW

File: lua/neotest-vstest/init.lua [hunk 1]
```diff
@@ -143,26 +143,40 @@ local function create_adapter(config)
...
+          type = "file",
+          path = fullpath,
+          limit = 1,
+        })
+      end
+      local project_files = find_project_file(true)
+      if vim.tbl_isempty(project_files) then
+        -- The csproj file may be in a sub-folder
+        project_files = find_project_file(false)
+      end
```
Comment (Nsidorenco) [inline lua/neotest-vstest/init.lua]: I think we need to let this be two separate code flows.
When searching upwards we only want the first/nearest project we can find.

If we search downwards we want all possible projects to be returned. 
Consider if all projects are nested in a `root/src` folder. There are many projects but only one test project. If we only return one project from the downwards descend we might end up with a "non-test" project, which will cause the whole `root/src` folder to be filtered away.

The easiest is probably to search both upwards and downwards always

File: lua/neotest-vstest/init.lua [hunk 2]
```diff
@@ -130,6 +130,34 @@ local function create_adapter(config)
...
+  ---@return returns a list of reachable projects (either parent or subfolder projects)
+  local function projects_reachable_from(fullpath, root)
+    local parent_project = vim.fs.find(function(path, _)
+      return path:match("%.[cf]sproj$")
+    end, {
+      upward = true,
+      type = "file",
+      stop = root,
+      path = fullpath,
+      limit = 1,
```
Comment (Dynge) [inline lua/neotest-vstest/init.lua]: Ive made the assumption that a file can only be the child of exactly one parent project.

While for the subfolders i return all matches since it may not be all projects that are relevant for the solution.

File: lua/neotest-vstest/init.lua (hunk 1 above)
Comment (Dynge) [reply lua/neotest-vstest/init.lua]: Please see the updated PR 🙂

File: lua/neotest-vstest/init.lua [hunk 3]
```diff
@@ -130,6 +130,34 @@ local function create_adapter(config)
     return true
   end
 
+  ---@param fullpath the path of the file to start searching from
+  ---@param stop_at the parent directory to stop searching at
```
Comment (liskin) [inline lua/neotest-vstest/init.lua]:
```suggestion
  ---@param fullpath string the path of the file to start searching from
  ---@param stop_at string the parent directory to stop searching at
```

(https://luals.github.io/wiki/annotations/#param)

**Review by liskin — CHANGES_REQUESTED** [review]

File: lua/neotest-vstest/init.lua [hunk 4]
```diff
@@ -143,26 +171,26 @@ local function create_adapter(config)
...
       return vim
         .iter(solution_info and solution_info.projects or {})
         :map(function(proj)
           return proj.proj_file
         end)
-        :any(function(proj_file)
-          for _, file in ipairs(project_files) do
-            if vim.fs.normalize(proj_file) == vim.fs.normalize(file) then
+        :any(function(sln_project_file)
+          for _, project_file in ipairs(project_files) do
```
[the remark this answers is not in the document]
Comment (liskin) [reply lua/neotest-vstest/init.lua]: Changing this to `projects` fixes this, and makes neotest-vstest discover all our tests again. So I can confirm this fixes https://github.com/Nsidorenco/neotest-vstest/pull/43#issuecomment-3418726088

**Comment by liskin** [comment]
> It may be nothing but I can't help but feel that we are doing a lot of redundant searches here.

Seems like, yeah:

```
DEBUG | 2025-10-19T17:41:15Z+0100 | .../bundle/start/neotest-vstest/lua/neotest-vstest/init.lua:188 | neotest-vstest: file 'tools' is not a part of the solution '<redacted>.sln'
DEBUG | 2025-10-19T17:41:15Z+0100 | .../bundle/start/neotest-vstest/lua/neotest-vstest/init.lua:188 | neotest-vstest: file 'tools' is not a part of the solution '<redacted>.sln'
DEBUG | 2025-10-19T17:41:15Z+0100 | .../bundle/start/neotest-vstest/lua/neotest-vstest/init.lua:188 | neotest-vstest: file 'tools' is not a part of the solution '<redacted>.sln'
DEBUG | 2025-10-19T17:41:15Z+0100 | .../bundle/start/neotest-vstest/lua/neotest-vstest/init.lua:188 | neotest-vstest: file 'tools' is not a part of the solution '<redacted>.sln'
DEBUG | 2025-10-19T17:41:15Z+0100 | .../bundle/start/neotest-vstest/lua/neotest-vstest/init.lua:188 | neotest-vstest: file 'tools' is not a part of the solution '<redacted>.sln'
DEBUG | 2025-10-19T17:41:15Z+0100 | .../bundle/start/neotest-vstest/lua/neotest-vstest/init.lua:188 | neotest-vstest: file 'tools' is not a part of the solution '<redacted>.sln'
DEBUG | 2025-10-19T17:41:15Z+0100 | .../bundle/start/neotest-vstest/lua/neotest-vstest/init.lua:188 | neotest-vstest: file 'tools' is not a part of the solution '<redacted>.sln'
DEBUG | 2025-10-19T17:41:15Z+0100 | .../bundle/start/neotest-vstest/lua/neotest-vstest/init.lua:188 | neotest-vstest: file 'tools' is not a part of the solution '<redacted>.sln'
DEBUG | 2025-10-19T17:41:15Z+0100 | .../bundle/start/neotest-vstest/lua/neotest-vstest/init.lua:188 | neotest-vstest: file 'tools' is not a part of the solution '<redacted>.sln'
DEBUG | 2025-10-19T17:41:15Z+0100 | .../bundle/start/neotest-vstest/lua/neotest-vstest/init.lua:188 | neotest-vstest: file 'tools' is not a part of the solution '<redacted>.sln'
DEBUG | 2025-10-19T17:41:15Z+0100 | .../bundle/start/neotest-vstest/lua/neotest-vstest/init.lua:188 | neotest-vstest: file 'tools' is not a part of the solution '<redacted>.sln'
DEBUG | 2025-10-19T17:41:15Z+0100 | .../bundle/start/neotest-vstest/lua/neotest-vstest/init.lua:188 | neotest-vstest: file 'tools' is not a part of the solution '<redacted>.sln'
DEBUG | 2025-10-19T17:41:15Z+0100 | .../bundle/start/neotest-vstest/lua/neotest-vstest/init.lua:188 | neotest-vstest: file 'tools' is not a part of the solution '<redacted>.sln'
DEBUG | 2025-10-19T17:41:15Z+0100 | .../bundle/start/neotest-vstest/lua/neotest-vstest/init.lua:188 | neotest-vstest: file 'tools' is not a part of the solution '<redacted>.sln'
```

I get a lot of this :-)

**Comment by Dynge** [comment]
> > It may be nothing but I can't help but feel that we are doing a lot of redundant searches here.
> 
> Agreed. My first implementation of this was actually not to recursively scan for project files but rather just get the list of all projects in the solution and then check if the direction we are currently filtering is a subpath of any project path.
> 
> The problem back then was that neovim had no real utilities for comparing paths and window-style pathing just have so many edge-cases.
> 
> `vim.fs.relpath` was added in neovim 0.11 so I'd be interested to see if we could figure out a solution that does not involve scanning all the directories, perhaps at the cost of bumping the minimal required neovim version of the plugin.

Ive just updated the PR to use string matching.

The point is that if i have path `a/b/c` and i need to know if I can filter it away or not, i need to check if `a/b/c` is either a project is a subdir of `a/b/c` or if `a/b/c` is a subdir of a project.

```
ProjectDir  DirUnderTest    
a/b         a/b/c           ProjectDir contained in DirUnderTest
a/b/c/d     a/b/c           DirUnderTest contained in ProjectDir
```

**Comment by liskin** [comment]
> `vim.fs.relpath` was added in neovim 0.11 so I'd be interested to see if we could figure out a solution that does not involve scanning all the directories, perhaps at the cost of bumping the minimal required neovim version of the plugin.

I tested the latest (2f5ad848fd1d8b7a04ed7c3b15e404d6bff6ffe5) version of this PR with
```lua
return vim.fs.relpath(fullpath, project_dir) ~= nil or vim.fs.relpath(project_dir, fullpath) ~= nil
```
instead of the
```lua
if fullpath:find(project_dir, 1, true) or project_dir:find(fullpath, 1, true) then
  return true
end
return false
```
and it seems to work fine. So if you're willing to bump the min to nvim 0.11, that's the solution I'd use

**Comment by liskin** [comment]
If, for whatever reason, we want to keep using the string match, then there's `vim.startswith` that has been there for years. Probably a bit more reliable than `string.find` which returns non-`nil` whenever the needle is anywhere in the haystack.

**Comment by Dynge** [comment]
> If, for whatever reason, we want to keep using the string match, then there's `vim.startswith` that has been there for years. Probably a bit more reliable than `string.find` which returns non-`nil` whenever the needle is anywhere in the haystack.

I kind of like the string matching. I cant think of any problematic edge-cases off the top of my head.
I like using startswith instead, its a little clearer what we are actually looking for :)

What do you think @Nsidorenco ?

**Comment by liskin** [comment]
> I kind of like the string matching. I cant think of any problematic edge-cases off the top of my head. I like using startswith instead, its a little clearer what we are actually looking for :)

Well, `/ab/cd/ef` starts with `/ab/c`, even though neither is a parent of the other, so that's one edge-case. Not a contrived one, either, as "test-harness" starts with "test". You could solve that by adding `/`, but then we're back at what @Nsidorenco said earlier:

> The problem back then was that neovim had no real utilities for comparing paths and window-style pathing just have so many edge-cases.

(I understood that as meaning "Windows-style" rather than "window-style".)

**Comment by liskin** [comment]
> Did any one of you use windows when testing out the functionality? I don't have a windows machine myself.

No, just Linux. I have a Windows VM somewhere but I don't even know how one installs neovim there so it'd take me more time than I have to go and try. I think it should be safe to assume relpath does the right thing. If not someone will file an issue sooner or later 🙂

**Comment by Nsidorenco** [comment]
Is it still necessary to use `vim.fs.normalize` after the change? Otherwise I think we’re good to go

File: lua/neotest-vstest/init.lua [hunk 5]
```diff
@@ -143,28 +143,27 @@ local function create_adapter(config)
...
+          -- path should be a subdir of a solution project
+          -- or the solution project should be a subdir of the path
+          return vim.fs.relpath(project_dir, fullpath) or vim.fs.relpath(fullpath, project_dir)
         end)
+      if not is_project_reachable then
+        logger.debug(
+          string.format(
+            "neotest-vstest: file '%s' is not a part of the solution '%s'",
+            rel_path,
+            solution_info.solution_file
```
[the remark this answers is not in the document]
Comment (Dynge) [reply lua/neotest-vstest/init.lua]: Yeah. I agree that there are some type checking errors, however I think a change to this would be out of scope for this PR :)

Could be made in another small refactoring PR.
